sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,1118 @@
+import so3g
+from spt3g import core
+import itertools
+import numpy as np
+from scipy.interpolate import interp1d
+from scipy.signal import convolve
+from tqdm.auto import tqdm
+import os
+import yaml
+import logging
+import sys
+import shutil
+from sotodlib.site_pipeline.util import init_logger
+
+
+log = logging.getLogger('bookbinder')
+if not log.hasHandlers():
+    init_logger('bookbinder')
+
+
+def setup_logger(logfile=None):
+    """
+    This setups up a logger for bookbinder. If a logfile is passed, it will
+    write to that file as well as stdout. It is useful to create a one-off
+    logger here instead of using `getLogger` because it allows us to set
+    a separate log-file for each bookbinder instance.
+    """
+    fmt = '%(asctime)s: %(message)s (%(levelname)s)'
+    log = logging.Logger('bookbinder', level=logging.DEBUG)
+
+    ch = logging.StreamHandler(sys.stdout)
+    formatter = logging.Formatter(fmt)
+    ch.setLevel(logging.INFO)
+    ch.setFormatter(formatter)
+    log.addHandler(ch)
+
+    if logfile is not None:
+        ch = logging.FileHandler(logfile)
+        ch.setLevel(logging.DEBUG)
+        ch.setFormatter(formatter)
+        log.addHandler(ch)
+
+    return log
+
+
+def get_frame_iter(files):
+    """
+    Returns a continuous iterator over frames for a list of files.
+    """
+    return itertools.chain(*[core.G3File(f) for f in files])
+
+
+def close_writer(writer):
+    """
+    Closes out a G3FileWriter with an end-processing frame. If None is passed,
+    this will not do anything.
+    """
+    if writer is None:
+        return
+    writer(core.G3Frame(core.G3FrameType.EndProcessing))
+
+
+def next_scan(it):
+    """
+    Returns the next Scan frame, along with any intermediate frames for an
+    iterator.
+    """
+    interm_frames = []
+    for frame in it:
+        if frame.type == core.G3FrameType.Scan:
+            return frame, interm_frames
+        interm_frames.append(frame)
+    return None, interm_frames
+
+
+class HKBlock:
+    def __init__(self, name):
+        self.name = name
+        self.data = {}
+        self.times = []
+    
+    def process_frame(self, frame):
+        if self.name not in frame['block_names']:
+            return
+        idx = list(frame['block_names']).index(self.name)
+        block = frame['blocks'][idx]
+
+        self.times.append(np.array(block.times) / core.G3Units.s)
+        for k, v in block.items():
+            if k not in self.data:
+                self.data[k] = []
+            self.data[k].append(v)
+    
+    def finalize(self):
+        if self.times:
+            self.times = np.hstack(self.times)
+            for k, v in self.data.items():
+                self.data[k] = np.hstack(v)
+        else:
+            self.times = np.array([], dtype=np.float32)
```

Comment:
Is float32 consistent with the times defined elsewhere in `finalize` and `process_frame`—i.e., in other frames (I believe numpy defaults to float64)—or does this not matter?